tsl/src/chunk.c

```diff
@@ -1350,43 +1455,660 @@ relation_split_info_free(RelationSplitInfo *rsi, int ti_options)
+		}
+
+		/*
+		 * Route each decompressed tuple to its corresponding partition's
+		 * compressor.
+		 */
+		for (int i = 0; i < nrows; i++)
+		{
+			int routing_index = route_tuple(decompressor.decompressed_slots[i], scontext->sp);
+			Assert(routing_index == 0 || routing_index == 1);
+			RelationWriteState *rws = &scontext->rws[routing_index];
+			/*
+			 * Since we're splitting a segment, the new segments will be
+			 * ordered like the original segment. Also, there is no risk of
+			 * the segments getting too big since we are only making segments
+			 * smaller.
+			 */
+			row_compressor_append_ordered_slot(&rws->compressor,
+											   decompressor.decompressed_slots[i]);
+		}
+
+		row_decompressor_close(&decompressor);
+		scontext->rws_index = 0;
+
+		/*
+		 * Call this function again to return the sub-segments.
+		 */
+		return route_next_compressed_tuple(slot, scontext, routing_index);
+	}
+
+	/* Update tuple count stats for compressed data */
+	bool isnull;
+	Datum count = slot_getattr(slot, csp->attnum_count, &isnull);
+	scontext->rws[*routing_index].stats.tuples_in_segments += DatumGetInt32(count);
+
+	/*
+	 * The compressed tuple (segment) can be routed without splitting it.
+	 */
+	Assert(*routing_index >= 0 && *routing_index < scontext->split_factor);
+	scontext->rws_index = scontext->split_factor;
+
+	return ExecFetchSlotHeapTuple(slot, false, NULL);
+}
+
+static double
+copy_tuples_for_split(SplitContext *scontext)
+{
+	Relation srcrel = scontext->rel;
+	TupleTableSlot *srcslot;
+	MemoryContext oldcxt;
+	EState *estate;
+	ExprContext *econtext;
+	TableScanDesc scan;
+	SplitPoint *sp = scontext->sp;
+
+	estate = CreateExecutorState();
+
+	/* Create the tuple slot */
+	srcslot = table_slot_create(srcrel, NULL);
+
+	/*
+	 * Scan through the rows using SnapshotAny to see everything so that we
+	 * can transfer tuples that are deleted or updated but still visible to
+	 * concurrent transactions.
+	 */
+	scan = table_beginscan(srcrel, SnapshotAny, 0, NULL);
+
+	/*
+	 * If a Hypercore TAM relation, we only want to read uncompressed data
+	 * since the compressed relation is split separately.
+	 */
+	hypercore_scan_set_skip_compressed(scan, true);
+
+	/*
+	 * Switch to per-tuple memory context and reset it for each tuple
+	 * produced, so we don't leak memory.
+	 */
+	econtext = GetPerTupleExprContext(estate);
+	oldcxt = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
+
+	/*
+	 * Read all the data from the split relation and route the tuples to the
+	 * new partitions. Do some vacuuming and cleanup at the same
+	 * time. Transfer all visibility information to the new relations.
+	 *
+	 * Main loop inspired by heapam_relation_copy_for_cluster() used to run
+	 * CLUSTER and VACUUM FULL on a table.
+	 */
+	double num_tuples = 0.0;
+	double tups_vacuumed = 0.0;
+	double tups_recently_dead = 0.0;
+
+	BufferHeapTupleTableSlot *hslot;
+	int routingindex = -1;
+
+	while (table_scan_getnextslot(scan, ForwardScanDirection, srcslot))
```

Comment:
Many branches there seem to miss coverage.

Reply:
Looks like only one "branch", which is dead tuple. This used to be covered, so something must have changed. I'll take a look.

Reply:
Should be fixed now.